File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return new ArrayList<List<Integer>>();
        }

        List<List<Integer>> levels = new ArrayList<>();
        Queue<TreeNode> bfs = new ArrayDeque<>();
        bfs.add(root);

        while (!bfs.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            int size = bfs.size();

            for (int i = 0; i < size; i++) {
                TreeNode currentNode = bfs.poll();
                level.add(currentNode.val);

                if (currentNode.left != null) {
                    bfs.add(currentNode.left);
                }
                if (currentNode.right != null) {
                    bfs.add(currentNode.right);
                }
            }

            levels.add(level);
        }

        return levels;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root, int left = INT_MIN, int right = INT_MAX) {
        if (!root) {
            return true;
        }
        else if ((left < root->val) && (root->val < right)) {
            return isValidBST(root->left, left, root->val) && isValidBST(root->right, root->val, right);
        }
        else {
            return false;
        }
    }
};
